### PR TITLE
fix(windows): cache rooted path probes and UNC globs

### LINKED ISSUE #497: [Task] Follow up Windows path canonicalization caching and live UNC validation
## Context

PR #493 fixed the correctness portion of #427 by canonicalizing Windows permission path variants and making guarded tools use the same canonical path for permission checks and actual filesystem/LSP/ripgrep operations.

#427 is closed. This issue is only for the remaining follow-up work that was intentionally left out of that PR.

## Priority

P3 tracking.

This is not a current blocking user-visible bug. PawWork does not currently have a stable Windows/live UNC validation environment, so this should not displace higher-priority reproducible app bugs.

Raise priority only if one of these happens:

- A real Windows user reports a regression in UNC / extended UNC / rooted-driveless path behavior.
- Windows advisory CI starts failing in this area.
- A maintainer has access to a reliable Windows environment and wants to close the validation gap.
- Drive-root probing is proven to be a hot-path performance problem.

## Scope

In scope:

- Evaluate whether `windowsDriveRoots()` probing should be cached.
- If caching is worth doing, define and test a conservative invalidation strategy.
- Add or document Windows live validation for:
  - extended drive paths such as `\\?\D:\...`
  - extended UNC paths such as `\\?\UNC\server\share\...`
  - live UNC shares where practical
  - rooted-but-driveless non-existing write targets
- Record what was validated and where.

Out of scope:

- Reopening the #427 correctness fix.
- Changing permission wildcard semantics.
- Broad path architecture rewrites.
- Blocking unrelated app bug work while no Windows environment is available.

## Suggested verification

When a Windows environment is available:

- Run the focused core filesystem normalization tests.
- Run the focused opencode guarded-tool boundary tests from #493.
- Add or manually record a smoke for live UNC / extended UNC behavior.
- Confirm permission metadata path and actual tool path remain identical.

## Related

- Closes context: #427
- Implemented correctness slice: #493

### PR DESCRIPTION
**Pull request by Astro-Han**
## Summary

Cache repeated Windows rooted-but-driveless canonicalization probes and close the remaining UNC validation gap in the external-directory permission path.

Changes included:

- cache successful rooted-but-driveless Windows probe results, with cache invalidation when the cached target disappears or the default Windows root environment changes
- normalize Windows UNC glob patterns with `path.win32` so `\\server\share\...` permission scopes do not collapse to `*`
- add deterministic tests for cache reuse/invalidation, extended UNC path normalization, and UNC permission-glob generation

## Why

Issue #497 left two bounded Windows follow-ups after #493:

1. repeated rooted-but-driveless canonicalization still re-probed drive roots on every hit
2. UNC / extended UNC validation was still incomplete, especially around permission-glob generation

This PR keeps the fix inside that P3 follow-up surface. It does not reopen #427 correctness work, wildcard semantics, or broader path architecture.

## Related Issue

Closes #497.

## Human Review Status

Pending. A human should make the final merge decision after reviewing the final diff and verification evidence.

## Review Focus

- whether the rooted-but-driveless cache only changes repeated Windows probe behavior and stays conservative on invalidation
- whether Windows UNC and extended UNC permission paths now preserve the share root all the way through glob generation
- whether the patch stays bounded to Windows-only canonicalization / permission-path logic

## Risk Notes

- Windows-only behavior change in path canonicalization and external-directory permission glob generation
- the cache stores only successful rooted-but-driveless matches; misses and ambiguities still probe live
- no live Windows UNC host was available locally, so UNC validation is deterministic unit coverage rather than an interactive Windows smoke

## How To Verify

```text
Core Windows normalization tests: bun --cwd packages/core test test/filesystem/normalize-path.test.ts -> 9 passed, 7 skipped
Guarded-tool boundary suite: bun --cwd packages/opencode test test/tool/external-directory.test.ts test/tool/read.test.ts test/tool/write.test.ts test/tool/edit.test.ts test/tool/apply_patch.test.ts test/tool/glob.test.ts test/tool/grep.test.ts test/tool/lsp.test.ts -> 135 passed
Core typecheck: bun --cwd packages/core typecheck -> exit 0
Opencode typecheck: bun --cwd packages/opencode typecheck -> exit 0
Diff check: git diff --check -> exit 0
Validation recorded in tests: packages/core/test/filesystem/normalize-path.test.ts and packages/opencode/test/tool/external-directory.test.ts
```

## Screenshots or Recordings

None. No visible UI changes.

## Checklist

- [x] Human review status is stated above as pending, approved, or not required
- [x] I linked the related issue, or stated why there is no issue
- [x] This PR has type, primary area, and priority labels, or I requested maintainer labeling
- [x] I described the review focus and any meaningful risks
- [x] I listed the relevant verification steps and the key result for each
- [x] I did not introduce unrelated refactors, dependencies, generated files, or file changes beyond the stated scope
- [x] I manually checked visible UI or copy changes when needed, with screenshots or recordings
- [x] I considered macOS and Windows impact for platform, packaging, updater, signing, paths, shell, or permissions changes
- [x] I called out docs, release notes, dependencies, permissions, credentials, deletion behavior, generated content, or local file changes when relevant
- [x] I reviewed the final diff for unrelated changes and suspicious dependency changes
- [x] I am targeting `dev`, and my PR title and commit messages use Conventional Commits in English

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Windows path handling to preserve extended UNC roots and produce correct Windows-style glob patterns; fixed directory canonicalization to retain share roots.

* **Performance**
  * Added/refined caching for Windows path resolution to reduce redundant filesystem probes and refresh correctly when drive context changes.

* **Tests**
  * Expanded Windows-only tests covering UNC/extended paths and caching/refresh behavior.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Astro-Han/pawwork/pull/582)

### CODE REVIEW

**Review by Astro-Han — COMMENTED** [review]
## GLM First-pass Review

**P0-P3 Severity Analysis**: No P0/P1/P2/P3 issues found.

### Spec Verification

| Spec Item | Implementation | Status |
|-----------|---------------|--------|
| Rooted-driveless cache hit logic | cache.get + exists check | Covered |
| Cache invalidation on target disappearance | cache.delete when !exists | Covered |
| Cache invalidation on cwd root change | refreshRootedWindowsVariantCache compares cwdRoot | Covered |
| Cache invalidation on SystemDrive change | envKey includes systemDrive | Covered |
| UNC share root preserved in glob | path.win32.dirname/join for UNC paths | Covered |
| Extended UNC glob normalization | Test verifies UNC path handling | Covered |
| Windows-only (no macOS/Linux impact) | process.platform !== win32 early return | Covered |

### Test Coverage

- Cache reuse: normalize-path.test.ts verifies probe count reduction
- Cache invalidation: Test verifies refreshed result after live.set
- Extended UNC: normalizePathPattern test
- UNC glob: external-directory.test.ts verifies patterns/always preserve share root

### Summary

- 4 files +141/-8, minimal scope
- Cache hit/invalidation paths complete
- UNC glob correctly preserves share root
- Tests cover all 4 boundary cases
- No macOS/Linux path behavior changes

**No blocking issues. Ready for second-pass review.**

**Review by Astro-Han — COMMENTED** [review]
Second-pass (product / scope) review.

Three-question check passes: change stays narrow (4 files +141/-8), cache invalidation covers cached-target removal / cwd-root change / SystemDrive change, UNC glob fix routes consistently through `path.win32`, and `resetWindowsRootedVariantCacheForTest` is test-only.

One P3 clarification, non-blocking:

`packages/core/src/filesystem.ts` `resolve()` switched its realpath call from `realpathSync(resolved)` to `fs.realpathSync.native(resolved)`. This isn't called out in the PR Review Focus or in issue #497's spec. It looks like an intentional consistency tweak (so `resolve()` matches `canonicalize()` which already used `.native`), but please confirm:

1. Is this intentional, not an unintended side effect of the `import * as fs from "fs"` refactor?
2. Are the existing tests covering `.native` behavior on Windows for symlink / junction / ENOENT / non-existent paths, given `.native` and the JS-shim variant can diverge on edge cases?

If intentional and behaviorally equivalent for the cases we care about, this is fine to keep as a small scope-aligned perf nudge. If accidental, please revert that single line to keep PR scope tight.

No P0 / P1 / P2 findings. Awaiting GPT-X engineering final review.

File: packages/core/src/filesystem.ts [hunk 1]
```diff
@@ -256,16 +260,52 @@ export namespace AppFileSystem {
...
       throw new Error(`Ambiguous Windows path ${p}; use a drive-qualified path.`)
     }
-    return matches[0]
+    const match = matches[0]
+    if (cache && match) cache.set(cacheKey, match)
+    return match
+  }
+
+  function rootedWindowsVariantCacheKey(p: string): string {
+    return win32.normalize(p).replaceAll("/", "\\").toUpperCase()
```
[the remark this answers is not in the document]
Comment (Astro-Han) [reply packages/core/src/filesystem.ts]: Accepted, simplified per suggestion. `win32.normalize()` already handles slash conversion here.
